### PR TITLE
Dynamic Graph first prototype

### PR DESCRIPTION
**Pull request by tonyyang-svail**
[Design Doc](https://github.com/tonyyang-svail/Paddle/blob/e40de7784119134c27033ce1f1d4b7a83a9f1e9b/paddle/contrib/dynamic/README.md)

### CODE REVIEW

File: paddle/contrib/dynamic/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wangkuiyi) [inline paddle/contrib/dynamic/CMakeLists.txt]: I think `tape` is a good namespace and directory name than `dynamic`, because `dynamic` is simply incomplete -- what we want to say is `dynamic_net`, which, however, is too long as a namespace name.

File: paddle/contrib/dynamic/README.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
+# Dynamic Graph on Fluid
+
+We can implement Dynet-like Tape(See this survey) by wrapping Paddle Fluid's `Operator`
```
Comment (wangkuiyi) [inline paddle/contrib/dynamic/README.md]: I feel that we need to add a sentence here to better explain the motivation of this effort:

> PaddlePaddle Fluid is targeting the autodiff without tape, which, however, is very challenging and we are still way from there. DyNet and PyTorch provide a good design idea, the *tape*, that significantly eases the challenge.  Also, DyNet provides a C++ API that is as convenient as Python but with higher efficiency and could conveniently integrate with industrial/production systems. This package, `tape`, combines the good of 
>
> 1. tape from PyTorch and DyNet
> 2. C++ API and core from DyNet
> 3. rich set of operators from PaddlePaddle

File: paddle/contrib/dynamic/README.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+// Optimizer
+paddle::tape::SGD sgd(0.001);
+
+// Data Feeder
+paddle::tape::Fill data_feeder(...);
+VariableHandle input(new paddle::tape::Variable("input"));
+
+for (int i = 0; i < 2; ++i) {
+  reset_global_tape();
```
Comment (wangkuiyi) [inline paddle/contrib/dynamic/README.md]: I would suggest that we don't expose the operations on the "tape" to users. Here, `reset_global_tape` looks a too low level API that should be out of the reach of end-users. 

I would suggest that we do:

```c++
for (int i = 0; i < ITERATIONS; ++i) {
  FORWARD_PASS() {
    ...
  }
}
```

to replace the call to `reset_global_tape`.

File: paddle/contrib/dynamic/README.md [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+  data_feeder(input);
+
+  auto loss = mean(linear2(linear1(input))); // compile time InferShape & InferVarType
```
Comment (wangkuiyi) [inline paddle/contrib/dynamic/README.md]: Let us use a more realistic example here where the minibatch includes labels that the cost operator can take.

File: paddle/contrib/dynamic/README.md [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+  LOG(INFO) << loss.value(); // Run forward up to loss
+
+  // Run backward, store gradient of w at w->Grad()
+  get_global_tape.Backward(loss);
```
Comment (wangkuiyi) [inline paddle/contrib/dynamic/README.md]: `get_global_tape` is a verb phrase and should be the name of a function, instead of a variable.

Also, either variable or function would expose the too low-level detail. I feel that we can just call

```c++
paddle::tape::Backward(loss);
```

File: paddle/contrib/dynamic/README.md [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+  // Update w
+  sgd(linear1.Params());
```
Comment (wangkuiyi) [inline paddle/contrib/dynamic/README.md]: I am afraid that defining SGD as a functor reduces the readability. I feel that it is more straightforward to have

```c++
sgd.Update(linear1.Params());
```

or just

```c++
sgd.Update(linear1);
```
